Engine/game/savegame.cpp

```diff
@@ -0,0 +1,596 @@
+{
+    AStream in(File::OpenFileRead(filename));
+    if (!in.get())
+        return kSvgErr_FileNotFound;
+
+    // Skip MS Windows Vista rich media header
+    RICH_GAME_MEDIA_HEADER rich_media_header;
+    rich_media_header.ReadFromFile(in.get());
+
+    // Check saved game signature
+    String svg_sig = String::FromStreamCount(in.get(), SavegameSource::SignatureLen);
+    if (svg_sig.Compare(SavegameSource::Signature))
+        return kSvgErr_SignatureFailed;
+
+    String desc_text;
+    if (desc && elems == kSvgDesc_UserText)
+        desc_text.Read(in.get(), SavegameDescription::UserTextLen);
+    else
+        for (; in->ReadByte(); ); // skip until null terminator
+    SavegameVersion svg_ver = (SavegameVersion)in->ReadInt32();
+
+    // Check saved game format version
+    if (svg_ver < kSvgVersion_LowestSupported ||
+        svg_ver > kSvgVersion_Current)
+    {
+        return kSvgErr_FormatVersionNotSupported;
+    }
+
+    ABitmap image;
+    if (desc && elems == kSvgDesc_UserImage)
+        image.reset(RestoreSaveImage(in.get()));
+    else
+        SkipSaveImage(in.get());
+
+    String version_str = String::FromStream(in.get(), SavegameDescription::EngineVersionLen);
+    Version eng_version(version_str);
+    if (eng_version > EngineVersion ||
+        eng_version < SavedgameLowestBackwardCompatVersion)
+    {
+        // Engine version is either non-forward or non-backward compatible
+        return kSvgErr_IncompatibleEngine;
+    }
+    String main_file;
+    if (desc && elems == kSvgDesc_EnvInfo)
+        main_file.Read(in.get(), SavegameDescription::MainDataFileLen);
+    else
+        for (; in->ReadByte(); ); // skip until null terminator
+
+    if (src)
+    {
+        src->Filename = filename;
+        src->Version = svg_ver;
+        src->InputStream.reset(in.release());
+    }
+    if (desc)
+    {
+        if (elems == kSvgDesc_EnvInfo)
+        {
+            desc->EngineVersion = eng_version;
+            desc->MainDataFilename = main_file;
+        }
+        if (elems == kSvgDesc_UserText)
+            desc->UserText = desc_text;
+        if (elems == kSvgDesc_UserImage)
+            desc->UserImage.reset(image.release());
+    }
+    return kSvgErr_NoError;
+}
+
+SavegameError OpenSavegame(const String &filename, SavegameSource &src, SavegameDescription &desc, SavegameDescElem elems)
+{
+    return OpenSavegameBase(filename, &src, &desc, elems);
+}
+
+SavegameError OpenSavegame(const String &filename, SavegameDescription &desc, SavegameDescElem elems)
+{
+    return OpenSavegameBase(filename, NULL, &desc, elems);
+}
+
+void DoBeforeRestore(PreservedParams &pp)
+{
+    pp.SpeechVOX = play.want_speech;
+    pp.MusicVOX = play.separate_music_lib;
+
+    unload_old_room();
+    delete raw_saved_screen;
+    raw_saved_screen = NULL;
+    remove_screen_overlay(-1);
+    is_complete_overlay = 0;
+    is_text_overlay = 0;
+
+    // cleanup dynamic sprites
+    for (int i = 1; i < spriteset.elements; ++i)
```

Comment:
If I remember correctly sprite 0 was reserved to keep "invalid sprite" graphic. I do not remember how it is used though.